enum and define with the same name collide

### Input C/C++ Header

```C
enum {
        A =
#define A 0
                A,
} anenum;
```

### Bindgen Invokation

<!-- Place either the `bindgen::Builder` or the command line flags used here. -->

```Rust
bindgen::Builder::default()
    .header("input.h")
    .generate()
    .unwrap()
```


### Actual Results


```
       Fresh bindgen v0.23.1                                                                                                                                                                                                                                                                                                        Compiling math_dot_h v0.1.0 (file:///home/def/p/rust/math_dot_h)
     Running `rustc --crate-name math_dot_h src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=93892b9e7e7b815c -C extra-filename=-93892b9e7e7b815c --out-dir /home/def/p/rust/math_dot_h/target/debug/deps -L dependency=/home/def/p/rust/math_dot_h/target/debug/deps`
error[E0428]: a value named `A` has already been defined in this module
 --> /home/def/p/rust/math_dot_h/target/debug/build/math_dot_h-7f0a85db403a8610/out/bindings.rs:4:1
  |
3 | pub const A: ::std::os::raw::c_uint = 0;
  | ---------------------------------------- previous definition of `A` here
4 | pub const A: _bindgen_ty_1 = _bindgen_ty_1::A;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `A` already defined

error: aborting due to previous error

error: Could not compile `math_dot_h`.

Caused by:
  process didn't exit successfully: `rustc --crate-name math_dot_h src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=93892b9e7e7b815c -C extra-filename=-93892b9e7e7b815c --out-dir /home/def/p/rust/math_dot_h/target/debug/deps -L dependency=/home/def/p/rust/math_dot_h/target/debug/deps` (exit code: 101)
```

### Expected Results

No compilation error, both the enum and the define have their own name space.

This is problematic IRL because `math.h` uses this construct to define `FP_*` constants:
https://sourceware.org/git/?p=glibc.git;a=blob;f=math/math.h;h=cfaed0ed98013830421afdf1ba6f13c01702831d;hb=HEAD#l318

```c
enum                                                                                                                                                                                                                                                                                                                               {                                                                                                                                                                                                                                                                                                                                  FP_NAN =                                                                                                                                                                                                                                                                                                                     # define FP_NAN 0                                                                                                                                                                                                                                                                                                                      FP_NAN,                                                                                                                                                                                                                                                                                                                        FP_INFINITE =                                                                                                                                                                                                                                                                                                                # define FP_INFINITE 1                                                                                                                                                                                                                                                                                                                 FP_INFINITE,
    FP_ZERO =
# define FP_ZERO 2
      FP_ZERO,
    FP_SUBNORMAL =
# define FP_SUBNORMAL 3
      FP_SUBNORMAL,
    FP_NORMAL =
# define FP_NORMAL 4
      FP_NORMAL
  };
```

### `RUST_LOG=bindgen` Output

<details>

```
   Compiling math_dot_h v0.1.0 (file:///home/def/p/rust/math_dot_h)
error[E0428]: a value named `A` has already been defined in this module
 --> /home/def/p/rust/math_dot_h/target/debug/build/math_dot_h-7f0a85db403a8610/out/bindings.rs:4:1
  |
3 | pub const A: ::std::os::raw::c_uint = 0;
  | ---------------------------------------- previous definition of `A` here
4 | pub const A: _bindgen_ty_1 = _bindgen_ty_1::A;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `A` already defined

error: aborting due to previous error

error: Could not compile `math_dot_h`.

To learn more, run the command again with --verbose.
```

</details>

I've started seeing this behavior on bindgen 0.26. While bindgen 0.25 generated code without errors. Bindgen 0.26 started giving me errors:
```
error[E0428]: the name `FP_NAN` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2820:1
     |
491  | pub const FP_NAN: libc::c_int = 0;
     | ---------------------------------- previous definition of the value `FP_NAN` here
...
2820 | pub const FP_NAN: _bindgen_ty_1 = _bindgen_ty_1::FP_NAN;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NAN` redefined here
     |
     = note: `FP_NAN` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_INFINITE` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2821:1
     |
492  | pub const FP_INFINITE: libc::c_int = 1;
     | --------------------------------------- previous definition of the value `FP_INFINITE` here
...
2821 | pub const FP_INFINITE: _bindgen_ty_1 = _bindgen_ty_1::FP_INFINITE;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_INFINITE` redefined here
     |
     = note: `FP_INFINITE` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_ZERO` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2822:1
     |
493  | pub const FP_ZERO: libc::c_int = 2;
     | ----------------------------------- previous definition of the value `FP_ZERO` here
...
2822 | pub const FP_ZERO: _bindgen_ty_1 = _bindgen_ty_1::FP_ZERO;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_ZERO` redefined here
     |
     = note: `FP_ZERO` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_SUBNORMAL` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2823:1
     |
494  | pub const FP_SUBNORMAL: libc::c_int = 3;
     | ---------------------------------------- previous definition of the value `FP_SUBNORMAL` here
...
2823 | pub const FP_SUBNORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_SUBNORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_SUBNORMAL` redefined here
     |
     = note: `FP_SUBNORMAL` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_NORMAL` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2824:1
     |
495  | pub const FP_NORMAL: libc::c_int = 4;
     | ------------------------------------- previous definition of the value `FP_NORMAL` here
...
2824 | pub const FP_NORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_NORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NORMAL` redefined here
     |
     = note: `FP_NORMAL` must be defined only once in the value namespace of this module

error: aborting due to 5 previous errors

error: Could not compile `ffmpeg-sys`.
```

As a workaround I'm just ignoring those types:
```
        .hide_type("FP_NAN")
        .hide_type("FP_INFINITE")
        .hide_type("FP_ZERO")
        .hide_type("FP_SUBNORMAL")
        .hide_type("FP_NORMAL")
```

I've encountered the same error with `bindgen-0.25.5` and `clang-4.0.0` on linux.
Crate [`magick-rust`](https://github.com/nlfiedler/magick-rust).
```
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4825:1
     |
747  | pub const FP_NAN: libc::c_uint = 0;
     | ----------------------------------- previous definition of the value `FP_NAN` here
...
4825 | pub const FP_NAN: _bindgen_ty_1 = _bindgen_ty_1::FP_NAN;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NAN` redefined here
     |
     = note: `FP_NAN` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_INFINITE` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4826:1
     |
748  | pub const FP_INFINITE: libc::c_uint = 1;
     | ---------------------------------------- previous definition of the value `FP_INFINITE` here
...
4826 | pub const FP_INFINITE: _bindgen_ty_1 = _bindgen_ty_1::FP_INFINITE;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_INFINITE` redefined here
     |
     = note: `FP_INFINITE` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_ZERO` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4827:1
     |
749  | pub const FP_ZERO: libc::c_uint = 2;
     | ------------------------------------ previous definition of the value `FP_ZERO` here
...
4827 | pub const FP_ZERO: _bindgen_ty_1 = _bindgen_ty_1::FP_ZERO;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_ZERO` redefined here
     |
     = note: `FP_ZERO` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_SUBNORMAL` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4828:1
     |
750  | pub const FP_SUBNORMAL: libc::c_uint = 3;
     | ----------------------------------------- previous definition of the value `FP_SUBNORMAL` here
...
4828 | pub const FP_SUBNORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_SUBNORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_SUBNORMAL` redefined here
     |
     = note: `FP_SUBNORMAL` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_NORMAL` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4829:1
     |
751  | pub const FP_NORMAL: libc::c_uint = 4;
     | -------------------------------------- previous definition of the value `FP_NORMAL` here
...
4829 | pub const FP_NORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_NORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NORMAL` redefined here
     |
     = note: `FP_NORMAL` must be defined only once in the value namespace of this module

```

--- a/bindgen-integration/build.rs
+++ b/bindgen-integration/build.rs
@@ -6,7 +6,7 @@ use std::env;
 use std::path::PathBuf;
 use std::sync::{Arc, RwLock};
 use bindgen::Builder;
-use bindgen::callbacks::ParseCallbacks;
+use bindgen::callbacks::{MacroParsingBehavior, ParseCallbacks};
 
 #[derive(Debug)]
 struct MacroCallback {
@@ -14,8 +14,14 @@ struct MacroCallback {
 }
 
 impl ParseCallbacks for MacroCallback {
-    fn parsed_macro(&self, _name: &str) {
-        self.macros.write().unwrap().insert(String::from(_name));
+    fn will_parse_macro(&self, name: &str) -> MacroParsingBehavior {
+        self.macros.write().unwrap().insert(name.into());
+
+        if name == "MY_ANNOYING_MACRO" {
+            return MacroParsingBehavior::Ignore
+        }
+
+        MacroParsingBehavior::Default
     }
 }
 

--- a/bindgen-integration/cpp/Test.h
+++ b/bindgen-integration/cpp/Test.h
@@ -2,6 +2,12 @@
 
 #define TESTMACRO
 
+enum {
+  MY_ANNOYING_MACRO =
+#define MY_ANNOYING_MACRO 1
+    MY_ANNOYING_MACRO,
+};
+
 class Test {
   int m_int;
   double m_double;

--- a/src/callbacks.rs
+++ b/src/callbacks.rs
@@ -5,11 +5,29 @@ pub use ir::int::IntKind;
 use std::fmt;
 use std::panic::UnwindSafe;
 
+/// An enum to allow ignoring parsing of macros.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum MacroParsingBehavior {
+    /// Ignore the macro, generating no code for it, or anything that depends on
+    /// it.
+    Ignore,
+    /// The default behavior bindgen would have otherwise.
+    Default,
+}
+
+impl Default for MacroParsingBehavior {
+    fn default() -> Self {
+        MacroParsingBehavior::Default
+    }
+}
+
 /// A trait to allow configuring different kinds of types in different
 /// situations.
 pub trait ParseCallbacks: fmt::Debug + UnwindSafe {
-    /// This function will be run on every macro that is identified
-    fn parsed_macro(&self, _name: &str) {}
+    /// This function will be run on every macro that is identified.
+    fn will_parse_macro(&self, _name: &str) -> MacroParsingBehavior {
+        MacroParsingBehavior::Default
+    }
 
     /// The integer kind an integer macro should have, given a name and the
     /// value of that macro, or `None` if you want the default to be chosen.

--- a/src/ir/var.rs
+++ b/src/ir/var.rs
@@ -1,5 +1,6 @@
 //! Intermediate representation of variables.
 
+use callbacks::MacroParsingBehavior;
 use super::context::{BindgenContext, TypeId};
 use super::dot::DotAttributes;
 use super::function::cursor_mangling;
@@ -122,9 +123,13 @@ impl ClangSubItemParser for Var {
         use cexpr::literal::CChar;
         match cursor.kind() {
             CXCursor_MacroDefinition => {
-
-                if let Some(visitor) = ctx.parse_callbacks() {
-                    visitor.parsed_macro(&cursor.spelling());
+                if let Some(callbacks) = ctx.parse_callbacks() {
+                    match callbacks.will_parse_macro(&cursor.spelling()) {
+                        MacroParsingBehavior::Ignore => {
+                            return Err(ParseError::Continue);
+                        }
+                        MacroParsingBehavior::Default => {}
+                    }
                 }
 
                 let value = parse_macro(ctx, &cursor);

